src/theory/evaluator.cpp

```diff
@@ -190,11 +281,12 @@ EvalResult Evaluator::evalInternal(TNode n,
 
         // Lambdas are evaluated in a recursive fashion because each evaluation
         // requires different substitutions
-        results[currNode] = evalInternal(op[1], lambdaArgs, lambdaVals);
+        results[currNode] =
+            evalInternal(op[1], lambdaArgs, lambdaVals, evalAsNode);
         if (results[currNode].d_tag == EvalResult::INVALID)
         {
-          // evaluation was invalid, we fail
-          return results[currNode];
+          // evaluation was invalid, we take the node of op[1] as the result
+          evalAsNode[currNode] = evalAsNode[op[1]];
```

Comment:
I think that there is an issue here if we have the same lambda twice invoked with different arguments and both evaluations are invalid. Maybe `evalAsNode` should be a `CDHashMap` and we push/pop the context around evaluating lambdas??

Reply:
Yes, good catch. I'll fix.